fvcbm.1
.\" -*- nroff -*-
.TH FVCBM 1 "2025-11-18" "fvcbm Version 3.3"
.SH NAME
fvcbm \- list directories of Commodore 64/128 compatible archive files
.SH SYNOPSIS
.B fvcbm
[
.B \-h
]
[
.B \-d
]
.B filename1
[
.IR filename2 ,
.IR \|.\|.\|.\|filenameN
]
.SH DESCRIPTION
.B fvcbm
displays the directories of Commodore 64 and 128 archive and self-dissolving
archive files, as well as raw floppy disk images.
It can be used to determine what is contained in an archive and how much disk
space it will take when uncompressed.
It is especially useful for previewing files on a network-connected machine
used for downloading Commodore software, and for viewing disk images outside
a Commodore 64 emulator.
.LP
If a given file cannot be found, several standard archive file extensions are
appended to the name in an attempt to find the desired file.
If a file named
.\" Some nroff's don't like: .IR "" ` \- '
`\-'
is given,
.B fvcbm
will attempt to read the standard input for an archive's contents.
Under Linux,
.B fvcbm
can read 1581 disks by specifying the floppy disk device (e.g.
.IR /dev/fd0 ).
The command
.B setfdprm
must first be used to specify that the drive contains a 1581 disk.
.LP
The default (wide) directory display contains a line for each file in the
archive. Each entry's display includes the file name, file type,
uncompressed length in bytes, uncompressed length in Commodore 254
byte disk blocks, compression method used, compression savings factor
(as a percentage reduction of the uncompressed file), and checksum.
The archive file name is displayed before the entries as well as the archive
title, if it exists (only some types support this).
.LP
Most columns are totalled in the last line.  Values displayed are the number
of files in the archive, total number of data bytes in the archive, number of
disk blocks required to hold all the files in the archive (not including
archive overhead), the archive type and version number (if known), overall
compression savings factor (not including the extraction program for
self-extracting files), and the number of disk blocks of compressed data
in the archive.
For self-extracting archives, a `+' and
the number of disk blocks taken up by the extraction program are displayed
following the number of blocks of data.  Adding the numbers
before and after the plus should roughly equal the size of the archive
file on disk (because of padding and rounding, the actual size on disk
may be slightly different).
.LP
The
.B \-d
option selects a style of directory output which is similar to that produced
by Commodore disk drives. Only the file name, size in blocks and
type are displayed for each entry in the archive. The summary line totals
the number of blocks used by the entries.
.LP
.B fvcbm
supports the following archive types: ARC230 (not to be confused with SEA
ARC), self-extracting ARC230 (SDA), Lynx, CS-DOS (LZH), self-extracting
CS-DOS (SFX), LBR (not to be confused with LBR for CP/M), 64Net files
(N64), PC64 emulator files (R/S/U/P00), emulator tape images (T64 and TAP) and
emulator disk images (D64 and X64 and variants), including 1541, 1571, 1581,
8050 and 8250 disk types (subdirectories a.k.a. partitions are not currently
supported). Note that there are several different TAP formats for different
computer types (Commodore and non-Commodore) but
.B fvcbm
supports only the Commodore variety.
.SH OPTIONS
.TP
.B \-h
Print a help message.
.TP
.B \-d
Display directory in Commodore disk directory format.
.TP
.B \-\-
Ends the list of options; only file names occur after this.
.SH "EXIT STATUS"
.B fvcbm
returns the following exit status:
.TP
.B 0
if no errors were encountered
.TP
.B 1
if
.B fvcbm
displayed the help message and exited
.TP
.B 2
if a file could not be found, or a read error occurred
.TP
.B 3
if the file type was not supported
.LP
If more than one filename caused an error, the status will reflect the
last error encountered.
.SH AUTHOR
Daniel Fandrich <dan@coneharvesters.com>
.LP
.B fvcbm
was inspired by Vernon D. Buerg's program
.I fv
for
displaying directories of MS-DOS archives.
.LP
Thanks to Guntram Blohm, Kevin Brisley, Teemu Rantanen, Haruyasu Yoshizaki and
Sven Petersen for their reference source code and/or documentation for various
archive formats.  Thanks also to Jouko Valta for beta testing and Tormod
Tjaberg for submitting improvements.
.LP
Feel free to contact the author if you have questions or suggestions for
improvement of the program.

.SH BUGS
Newer versions of some archivers may create files which are not recognized by
.B fvcbm
(please contact the author if you find one).
.LP
Reading an archive from the standard input does not work for some archive
types in some environments; when it does work, the file lengths can be
incorrect for Lynx and PC64 files.
.LP
`Locked' and `Not Closed' file status for D64 and X64 archive types is not
displayed.
.LP
Some archive types (especially D64 and ARC) can occasionally be misidentified.
This should not happen if the files are using one of the standard file
extensions for the type.
.LP
Older versions of Lynx (at least versions IX
and below) as well as PC64 file types like .P00, etc. files do not explicitly
store the length of the final file in the archive.  This last file can
therefore be improperly lengthened if it is transferred using a file transfer
protocol like XMODEM which adds extraneous characters to the end of transferred
files, or is transferred to an OS that does not preserve exact file lengths,
like CP/M 2.2.
.B fvcbm
(and Lynx itself) has no way of determining if this has happened, and will
display the incorrect file length if this is the case.
.SH COPYRIGHT
.B fvcbm
is copyright \(co 1995-2025 by Daniel Fandrich, et. al.
It is provided \(lqas is\(rq, without any express or implied warranties.
See the file COPYING, which should have been included with the program,
for details.
.SH "SEE ALSO"
.BR c1541 (1),
.BR cbmconvert (1),
.BR cbmfs (4),
.BR lha (1),
.BR lscbm (1),
.BR setfdprm (8),
.BR vice (1)


fvcbm.c
/*
 * fvcbm.c
 *
 * File View for Commodore 64 and 128 self dissolving and other archives
 * Displays a list of the file names and file info contained within an archive
 * Supports ARC230 (C64 & C128 versions), ARC230 SDA, Lynx, LHA, LHA SFX
 *   T64 (tape image), TAP (raw tape image), D64/X64 (disk image),
 *   PC64 (headered R/S/U/P00 file), 64Net (.N64) archive formats
 * Inspired by Vernon D. Buerg's FV program for MS-DOS archives
 *
 * Written under MS-DOS; compiled under Turbo C ver. 2.0; use -a- option to
 *   align structures on bytes in cbmarcs.c; link with WILDARGS.OBJ for
 *	 wildcard pattern matching
 * Microsoft C support; compile with -DMSC and -Zp1 option to align structures
 *   on bytes in cbmarcs.c (note: this support has not been properly tested)
 * Support for 386 SCO UNIX System V/386 Release 3.2; compile cbmarcs.c with
 *   -Zp1 option to align structures on bytes; compile fvcbm.c without -Zp1
 * Big-endian architecture support has been added but is untested
 *
 * Source file tab size is 4
*
 * fvcbm is copyright (C) 1995-2025 by Daniel Fandrich
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with fvcbm; if not, see <https://www.gnu.org/licenses/>.
 */

/******************************************************************************
* Include files
******************************************************************************/
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#if defined(__TURBOC__)
#include <dir.h>
#include <io.h>
#include <fcntl.h>

#elif defined(MSC) || defined(__ZTC__) || defined(__SC__) || defined(__WATCOMC__) || defined(__DJGPP__)
#include <io.h>
#include <fcntl.h>
#endif

/* Get some automatic filename globbing */
#ifdef __ZTC__
#undef MSDOS
#define MSDOS
#include <dos.h>
WILDCARDS
#endif

#include "cbmarcs.h"

/******************************************************************************
* Constants
******************************************************************************/
#define VERSION "3.4dev"
#define VERDATE "2025-11-18"

#if defined(__MSDOS__)
#define MAXPATH 80			/* length of longest permissible file path */
#else
#define MAXPATH 1025			/* length of longest permissible file path */
#endif

#if defined(__TURBOC__)
unsigned _stklen = 8000;	/* printf() does strange things sometimes with the
							   default 4k stack */
#endif

const char * const ProgName = "fvcbm";	/* should rather use argv[0] on Unix */

/* Archive extensions to try, in this order */
/* List must end with NULL */
#define MAX_EXT_LEN	4		/* longest extension in this list, including dot */
const char * const Extensions[] = {
".sda", ".sfx", ".d64", ".x64", ".t64", ".lnx", ".lzh",
".n64", ".arc", ".lbr", ".p00", ".s00", ".u00", ".r00",
".tap",
NULL
};

/******************************************************************************
* Global Variables
******************************************************************************/
int WideFormat;			/* zero when 1541-style listing is selected */

/******************************************************************************
* Display header information about an archive
******************************************************************************/
static void DisplayHeader(enum ArchiveTypes ArchiveType, const char *Name)
{
	(void) ArchiveType;
	if (WideFormat) {
		if (Name)
			printf("Title:   %s\n", Name);
		printf("\nName              Type  Length  Blks  Method     SF   Now   Check\n");
		printf(  "================  ====  ======  ====  ========  ====  ====  =====\n");

	} else {
		if (Name)
			printf("\n     \"%s\"", Name);
		printf("\n");
	}
}

/******************************************************************************
* Display a file's name and info
******************************************************************************/
static int DisplayFile(const char *Name, const char *Type, unsigned long Length,
		unsigned Blocks, const char *Storage, int Compression,
		unsigned BlocksNow, long Checksum)
{
	if (WideFormat) {
		printf("%-16s  %s  %7lu  %4u  %-8s %4d%%  %4u",
				Name, Type, Length, Blocks, Storage, Compression, BlocksNow);
		if (Checksum >= 0)
			printf("   %04X", (int) Checksum);
		printf("\n");
	} else {
		char QuoteName[19];
		QuoteName[0] = '"';
		strcpy(QuoteName+1,Name);
		strcat(QuoteName,"\"");
		printf("%-5u%-18s %s\n",
			Blocks, QuoteName, Type);
	}
	return 0;
}

/******************************************************************************
* Display summary information about an archive
******************************************************************************/
static void DisplayTrailer(enum ArchiveTypes ArchiveType, const struct ArcTotals *Totals)
{
	if (WideFormat) {
		printf("================  ====  ======  ====  ========  ====  ====  =====\n");
		printf("*total %5u           %7lu  %4d  %s",
			Totals->ArchiveEntries,
			Totals->TotalLength,
			Totals->TotalBlocks,
			ArchiveFormats[ArchiveType]);
		if (Totals->Version > 0)
			printf("%4u", Totals->Version);
		else if (Totals->Version < 0)
			printf("%2u.%u",
				-Totals->Version / 10,
				-Totals->Version - 10 * (-Totals->Version / 10)
			);
		else
			printf("    ");
		printf(" %4d%%  %4d",
			Totals->TotalBlocks == 0 ?
				0 :
				(unsigned) (100 - (Totals->TotalBlocksNow * 100L / (Totals->TotalBlocks))),
			Totals->TotalBlocksNow
		);
		if (Totals->DearcerBlocks > 0)
			printf("+%d\n", Totals->DearcerBlocks);
		else
			printf("\n");

	} else
		printf("%u BLOCKS USED.\n", Totals->TotalBlocks);
}


/******************************************************************************
* Main loop
******************************************************************************/
int main(int argc, char *argv[])
{
	int ArgNum;
	int Error = 0;
	int DispError;
	int FirstFileName = 1;
	char FileName[MAXPATH+1];
	FILE *InFile;
	enum ArchiveTypes ArchiveType;
	struct ArcTotals Totals;

#ifndef __Z88DK
	setvbuf(stdout, NULL, _IOLBF, 82);		/* speed up screen output */
#endif

	if ((argc > 1) &&
		((argv[FirstFileName][0] == '-') || (argv[FirstFileName][0] == '/')) &&
		((argv[FirstFileName][1] == 'd')
#ifdef CPM
		  || (argv[FirstFileName][1] == 'D')
#endif
		) && (argv[FirstFileName][2] == '\x0')) {
		WideFormat = 0;		/* 1541-style output */
		++FirstFileName;
	} else {
		WideFormat = 1;		/* wide FV-style output */
	}

	if ((argc <= FirstFileName) ||
		(((argv[FirstFileName][0] == '-') || (argv[FirstFileName][0] == '/')) &&
		 ((argv[FirstFileName][1] == '?') || (argv[FirstFileName][1] == 'h')
#ifdef CPM
		  || (argv[FirstFileName][1] == 'H')
#endif
		 ) &&
		 (argv[FirstFileName][2] == '\x0'))) {
		printf("%s  ver. " VERSION "  " VERDATE "  by Daniel Fandrich\n", ProgName);
		printf("Usage:\n  %s [-d] filename1 [filenameN ...]\n"
			   "View directory of Commodore 64/128 archive and self-dissolving archive files.\n"
			   "Supports ARC230, Lynx, LZH (SFX), T64, TAP, D64, X64, N64, PC64 & LBR archive\n"
			   "types.\n"
			   "fvcbm is copyright (C) 1995-2025 by Daniel Fandrich, et. al.\n"
			   "This program comes with NO WARRANTY. See the file COPYING for details.\n",
			   ProgName);
		return 1;
	}

	/* -- ends options */
	if ((argv[FirstFileName][0] == '-') && (argv[FirstFileName][1] == '-')) {
		++FirstFileName;
	}

/******************************************************************************
* Loop through archive display for each file name
******************************************************************************/
	for (ArgNum=FirstFileName; ArgNum<argc; ++ArgNum) {

/******************************************************************************
* Open the archive file
******************************************************************************/
		strncpy(FileName, argv[ArgNum], sizeof(FileName) - MAX_EXT_LEN);
		FileName[sizeof(FileName)-1] = 0;

		if (strcmp(FileName,"-") == 0) {		/* "-" means read from stdin */
			InFile = stdin;
#ifdef __MSDOS__
			setmode(fileno(stdin), O_BINARY);	/* put standard input into binary mode */
#endif
		} else if ((InFile = fopen(FileName, "rb")) == NULL) {

/******************************************************************************
* Given name wasn't found--add extensions and keep searching
******************************************************************************/
			const char * const *Ext;
			char TryFileName[MAXPATH+1];

			for (Ext = Extensions; *Ext != NULL; ++Ext) {
				strcat(strcpy(TryFileName, FileName), *Ext);
				if ((InFile = fopen(TryFileName, "rb")) != NULL) {
					strcpy(FileName, TryFileName);
					break;
				}
			}

/******************************************************************************
* Couldn't find any variation of the file name
******************************************************************************/
			if (InFile == NULL) {
				perror(FileName);
				printf("\n");
				Error = 2;
				continue;		/* go do next file */
			}
		}

/******************************************************************************
* Display header
* To do: Add display of archive comment
******************************************************************************/
		printf("Archive: %s\n", FileName);

		if ((ArchiveType = DetermineArchiveType(InFile,FileName)) == UnknownArchive) {
			fprintf(stderr,"%s: Not a known Commodore archive\n", ProgName);
			Error = 3;
		} else {

/******************************************************************************
* Display the archive contents
******************************************************************************/
			if ((DispError = DirArchive(InFile, ArchiveType, &Totals,
										DisplayHeader, DisplayFile)) != 0)
				Error = DispError;
			else
				DisplayTrailer(ArchiveType, &Totals);	/* show output trailer */
		}

/******************************************************************************
* Go do next file name on command line
******************************************************************************/
		fclose(InFile);
		if (ArgNum<argc-1)
			printf("\n");
	}

	fflush(stdout);		/* Make sure the buffered output is displayed */

	return Error;
}
